File: Trees1.java
//validate binary search tree
//Time Complexity O(N) and space complexity O(N)
class Solution {

    
    public boolean isValidBST(TreeNode root) {
      
        return helper(root,null,null);
    }

    public boolean helper(TreeNode root, Integer high, Integer low){

        if(root==null) return true;

        if(low!=null && root.val <= low) return false;

        if(high!=null && root.val >=high) return false;

        boolean left = helper(root.left,root.val,low);

        boolean right = helper(root.right,high,root.val);

        return left && right;
       
    }
}



//construct tree from inorder and preorder
//Time complexity O(N) and Space complexity O(N)
class Solution {
    int preorderIndx;
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        preorderIndx=0;
        HashMap<Integer,Integer> map = new HashMap<>();

        for(int i=0;i<inorder.length;i++){
            map.put(inorder[i],i);
        }

        return buildPath(map,0,inorder.length-1,preorder);
        
    }

    public TreeNode buildPath(HashMap<Integer,Integer> map, int left, int end, int[] preorder){

        if(left>end) return null;

        int rootVal = preorder[preorderIndx];
        preorderIndx++;
        TreeNode root = new TreeNode(rootVal);

        int rootIdx = map.get(rootVal);

        root.left = buildPath(map,left,rootIdx-1,preorder);
        root.right = buildPath(map,rootIdx+1,end,preorder);
        return root;
    }
}

File: trees1.py
# problem 1 : 98. Validate Binary Search Tree : https://leetcode.com/problems/validate-binary-search-tree/
# Time Complexity : O(n)
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : None

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        if not root:
            return True
            
        stack = [(root, float('-inf'), float('inf')), ] 
        while stack:
            root, lower, upper = stack.pop()
            if not root:
                continue
            val = root.val
            if val <= lower or val >= upper:
                return False
            stack.append((root.right, val, upper))
            stack.append((root.left, lower, val))
        return True  


# problem 2 : 105. Construct Binary Tree from Preorder and Inorder Traversal : https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/
# Time Complexity : O(n)
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : None

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        self.idx = 0
        self.in_index = {}
        for i in range(len(inorder)):
            self.in_index[inorder[i]] = i
        
        return self.helper(preorder, inorder, 0, len(inorder)-1)
        
    def helper(self, preorder, inorder, start, end):
        if start > end:
            return None
        ## logic
        index = self.in_index[preorder[self.idx]]
        root = TreeNode(preorder[self.idx])
        self.idx += 1
        root.left = self.helper(preorder, inorder, start, index-1)
        root.right = self.helper(preorder, inorder, index+1, end)
        return root